pkg/object/tcpproxy/connection.go

```diff
@@ -0,0 +1,395 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package tcpproxy
+
+import (
+	"errors"
+	"io"
+	"net"
+	"runtime/debug"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/megaease/easegress/pkg/logger"
+	"github.com/megaease/easegress/pkg/util/iobufferpool"
+	"github.com/megaease/easegress/pkg/util/limitlistener"
+	"github.com/megaease/easegress/pkg/util/timerpool"
+)
+
+const writeBufSize = 8
+
+var tcpBufferPool = sync.Pool{
+	New: func() interface{} {
+		return make([]byte, iobufferpool.DefaultBufferReadCapacity)
+	},
+}
+
+// Connection wrap tcp connection
+type Connection struct {
+	closed     uint32
+	rawConn    net.Conn
+	localAddr  net.Addr
+	remoteAddr net.Addr
+
+	lastBytesSizeRead  int
+	lastWriteSizeWrite int
+
+	readBuffer      []byte
+	writeBuffers    net.Buffers
+	ioBuffers       []*iobufferpool.StreamBuffer
+	writeBufferChan chan *iobufferpool.StreamBuffer
+
+	mu               sync.Mutex
+	startOnce        sync.Once
+	connStopChan     chan struct{} // use for connection close
+	listenerStopChan chan struct{} // use for listener close
+
+	onRead  func(buffer *iobufferpool.StreamBuffer) // execute read filters
+	onClose func(event ConnectionEvent)
+}
+
+// NewClientConn wrap connection create from client
+func NewClientConn(conn net.Conn, listenerStopChan chan struct{}) *Connection {
+	return &Connection{
+		rawConn:          conn,
+		localAddr:        conn.LocalAddr(),
+		remoteAddr:       conn.RemoteAddr(),
+		listenerStopChan: listenerStopChan,
+
+		mu:              sync.Mutex{},
+		connStopChan:    make(chan struct{}),
+		writeBufferChan: make(chan *iobufferpool.StreamBuffer, writeBufSize),
+	}
+}
+
+// SetOnRead set connection read handle
+func (c *Connection) SetOnRead(onRead func(buffer *iobufferpool.StreamBuffer)) {
+	c.onRead = onRead
+}
+
+// SetOnClose set close callback
+func (c *Connection) SetOnClose(onclose func(event ConnectionEvent)) {
+	c.onClose = onclose
+}
+
+// Start running connection read/write loop
+func (c *Connection) Start() {
+	c.startOnce.Do(func() {
+		c.goWithRecover(func() {
+			c.startReadLoop()
+		}, func(r interface{}) {
+			_ = c.Close(NoFlush, LocalClose)
+		})
+
+		c.goWithRecover(func() {
+			c.startWriteLoop()
+		}, func(r interface{}) {
+			_ = c.Close(NoFlush, LocalClose)
+		})
+	})
+}
+
+func (c *Connection) goWithRecover(handler func(), recoverHandler func(r interface{})) {
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logger.Errorf("tcp connection goroutine panic: %v\n%s\n", r, string(debug.Stack()))
+				if recoverHandler != nil {
+					// it is not needed to wrap recoverHandler with go func in the current scenario
+					recoverHandler(r)
+				}
+			}
+		}()
+		handler()
+	}()
+}
```

Comment:
It is good to use some patterns, but it is also important to keep code short, plain, and easy to read, we need to balance.

```suggestion
	fnRecover := func() {
		if r := recover(); r != nil {
			logger.Errorf("tcp connection goroutine panic: %v\n%s\n", r, string(debug.Stack()))
			c.Close(NoFlush, LocalClose)
		}
	}
		
	c.startOnce.Do(func() {
		go func() {
			defer fnRecover()
			c.startReadLoop()
		}()

		go func() {
			defer fnRecover()
			c.startWriteLoop()
		}()
	})
}
```